Add a SILGenCleanup pass and CanonicalizeInstruction utility.

atrick opened the pull request:
Canonicalize instruction will be a superset of
simplifyInstruction (once all the transforms are fixed for ownership
SIL). Additionally, it will also include simple SSA-based
canonicalization that requires new instruction creation. It may not
perform any optimization that interferes with diagnostics or increases
compile time.

The SILGenCleanup pass runs before diagnostics to perform any
canonicalization required by diagnostics.

Canonicalization replaces simplifyInstruction in SILCombine so we can
easily factor some existing SILCombine transforms into canonicalization.

atrick committed:
Canonicalize loads in SILGenCleanup.

Reimplement load instruction canonicalization as part of the
CanonicalizeInstruction utility.

gottesmm commented on include/swift/SILOptimizer/Utils/CanonicalizeInstruction.h:
```diff
@@ -0,0 +1,78 @@
+  SILBasicBlock::iterator canonicalize(SILInstruction *inst);
+
+  /// Record a newly generated instruction.
+  virtual void genInstruction(SILInstruction *inst) = 0;
```
notifyNewInstruction would be a clearer name.

gottesmm commented on include/swift/SILOptimizer/Utils/CanonicalizeInstruction.h:
```diff
@@ -0,0 +1,78 @@
+  virtual void killInstruction(SILInstruction *inst) = 0;
+
+  /// Record a SIL value that has acquired new users.
+  virtual void newUsers(SILValue value) = 0;
```
notifyHasNewUsers?

gottesmm commented on lib/SILOptimizer/Mandatory/SILGenCleanup.cpp:
```diff
@@ -0,0 +1,78 @@
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+// Perform peephole-style "cleanup" to aid SIL diagnostic passes.
```
Same here. This is not a proper doxygen file level comment.

gottesmm commented on lib/SILOptimizer/Mandatory/SILGenCleanup.cpp:
```diff
@@ -0,0 +1,78 @@
+using namespace swift;
+
+// Define a CanonicalizeInstruction subclass for use in SILGenCleanup.
+struct SILGenCanonicalize : CanonicalizeInstruction {
```
Can you make this a final inheritance and mark all of the overridden methods with override?

gottesmm commented on lib/SILOptimizer/Mandatory/SILGenCleanup.cpp:
```diff
@@ -0,0 +1,78 @@
+};
+
+//===----------------------------------------------------------------------===//
+//                            Top Level Pass
```
We generally have been using Top Level Entrypoint. Also, are you sure that Top Level Pass is centered? Here is an emacs macro that does this automagically if you need:

```
(defun center-banner ()
  "Create a centered banner using the data on the current line."
  (interactive)
  (let* ((slash-string (make-string 2 ?/))
         (equal-string (make-string 3 ?=))
         (banner-string (concat slash-string
                                equal-string
                                (make-string 70 ?-)
                                equal-string
                                slash-string)))
    (save-excursion
      ;; Create banner on previous line.
      (forward-line -1)
      (end-of-line)
      (newline)
      (insert banner-string)
      (forward-line 1)
      (let ((lm (current-left-margin))
            line-length)
        (beginning-of-line)
        (delete-horizontal-space)
        (end-of-line)
        (delete-horizontal-space)
        (setq line-length (current-column))
        (if (> (- fill-column lm line-length 2) 0)
            (indent-line-to
             (+ lm (- (/ (- fill-column lm line-length) 2) 2))))
        (beginning-of-line)
        (insert-before-markers slash-string)
        (end-of-line)
        (newline)
        (insert banner-string)))))
```

gottesmm commented on lib/SILOptimizer/SILCombiner/SILCombine.cpp:
```diff
@@ -113,6 +135,39 @@ void SILCombineWorklist::add(SILInstruction *I) {
 }
 
+// Define a CanonicalizeInstruction subclass for use in SILCombine.
+class SILCombineCanonicalize : CanonicalizeInstruction {
```
Can you make this final/add override to all of the overridden entry points.

gottesmm commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -0,0 +1,176 @@
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+// SSA-peephole transformations that yield a more canonical SIL representation.
```
Another file level header that needs to be doxygenified.

gottesmm commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -0,0 +1,176 @@
+STATISTIC(NumSimplified, "Number of instructions simplified");
+
+// Tracing within the implementation can also be activiated by the pass.
+#undef DEBUG_TYPE
```
Can you document here the header that we want to have the original DEBUG_TYPE?

atrick replied:
I don't know what you want but I did something more explicit.

atrick replied:
The new implementation was reviewed in person.

gottesmm reviewed (commented):
Quick drive by. I want to do another pass through on this before I give the LGTM. Also can you split this up so that:

1. Adding the new code/utilities is one commit.
2. We have separate commits for each pass that are being updated.

atrick reviewed (commented):
I will rebase this soon after splitting it into 4 commits.

atrick committed:
Cleanup replaceAllUsesAndErase, return an iterator, allow erase handlers.

This will make the forthcoming CanonicalizeInstruction interface more
clear.

This is generally the better approach to utilities that mutate the
instruction stream. It avoids the temptation to assume that only a
single instruction will be deleted or that only instructions before
the current iterator will be deleted. This often happens to work but
eventually fails in the presense of debug and end-of-scope
instructions.

A function returning an iterator has a more clear contract than one
accepting some iterator reference of unknown
providence. Unfortunately, it doesn't work at the lowest level of
utilities, such as recursivelyDeleteTriviallyDeadInstructions, where
we want to handle instruction batches.

atrick committed:
Add a CanonicalizeInstruction utility.

CanonicalizeInstruction will be a superset of
simplifyInstruction (once all the transforms are fixed for ownership
SIL). Additionally, it will also include simple SSA-based
canonicalization that requires new instruction creation. It may not
perform any optimization that interferes with diagnostics or increases
compile time.

Canonicalization replaces simplifyInstruction in SILCombine so we can
easily factor some existing SILCombine transforms into canonicalization.

atrick committed:
Add a SILGenCleanup pass.

The SILGenCleanup pass runs before diagnostics to perform any
canonicalization required by diagnostics.

atrick committed:
Fix IRGen/enum.sil RUN line.

Add -disable-diagnostic-passes because we want to test IRGen on
precisely the SIL input sequence.

atrick committed:
Fixup access_markers.sil test case.

Canonical access markers should always guard some memory
operation. It's valid to delete dead access markers even before
exclusivity diagnostics. Deleting dead memory operations before
diagnostics would *not* be ok.

atrick committed:
SILGenCleanup tests.

atrick committed:
Canonicalize nontrivial loads.

This adds support to the load->struct_extract canonicalization for
nontrivial element types which look like:

load [copy]
borrow
struct_extract
...uses...
end_borrow
destroy

atrick committed:
Extend an exclusivity diagnostic test for nontrivial let's.

atrick commented:
@gottesmm this PR now has the full set of commits that I want to get into 5.1 to fix exclusivity diagnostics. I'll make a separate PR for store canonicalization for master only.

atrick commented:
@swift-ci test

atrick commented:
@swift-ci test source compatibility

atrick commented:
@swift-ci benchmark

atrick commented:
@swift-ci test

atrick commented:
@swift-ci benchmark

atrick commented:
Forced pushed a minor test case fix for i386.

atrick commented:
@swift-ci test.

atrick commented:
@swift-ci benchmark.

atrick commented:
@swift-ci benchmark

atrick commented:
@swift-ci benchmark

atrick commented:
@gottesmm I don't think there are any more build/test issues to fix here, but I'll wait to restart CI until you've finished your review.

gottesmm commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -64,16 +95,121 @@ simplifyAndReplace(SILInstruction *inst, CanonicalizeInstruction &pass) {
+    Projection proj;
+    SingleValueInstruction *svi;
+
+    // When sorting, just look at the projection and ignore the instruction.
```
I know that this is just moving some stuff around, but can you add to the comment here saying that not doing this results in non-determinism since it is possible to have multiple projections that are the same with different instruction pointers. So when we sort, the order may change subtly between compiler invocations == unhappiness. I just want to make sure that this is memorialized in the comment to prevent someone from removing this due to lack of context.

atrick replied:
Yes, done.

gottesmm commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -64,16 +95,121 @@ simplifyAndReplace(SILInstruction *inst, CanonicalizeInstruction &pass) {
+    pass.notifyNewInstruction(projInst);
+
+    lastNewLoad = B.createLoad(loadInst->getLoc(), projInst,
+                               loadInst->getOwnershipQualifier());
```
Even though I think that today this is safe (since you are only handling load [trivial], load), this may be incorrect in future cases where we handle [copy], [take]. Specifically if we are shrinking a projection to just use a trivial part of it, then you want to use SILBuilder::createTrivialLoadOr(...). If I am correct, then I am fine if you either use that API or if you add an assert here with a comment saying how this code needs to be updated when this code handles load [copy], load [take].

atrick replied:
This supports copies in the next commit in the same PR. The SILBuilder utility doesn't help because the client code needs to know how the load will be qualified to generate destroys correctly (the utility looks like it can cause hidden bugs).

gottesmm commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -64,16 +95,121 @@ simplifyAndReplace(SILInstruction *inst, CanonicalizeInstruction &pass) {
+//   ->
+// (load (struct_element_addr %base, #field)
+static SILBasicBlock::iterator
+splitAggregateLoad(LoadInst *loadInst, CanonicalizeInstruction &pass) {
```
Should this also handle load_borrow? I am not sure what that would mean in terms of the structuring. Just some food for thought.

atrick replied:
Possibly, depending on where SILGen emits LoadBorrow. But not in this 5.1 PR. I added a TODO here.

gottesmm commented on include/swift/SILOptimizer/Utils/CanonicalizeInstruction.h:
```diff
@@ -0,0 +1,87 @@
+  /// notifyNewInstruction(), killInstruction(), and notifyNewUsers().
+  ///
+  /// Warning: If the \p inst argument is killed and the client immediately
+  /// erases it, then it may be an invalid pointer upon return.
```
Can you replace the 2nd it here with `\p inst`. Then it is really clear that inst may be an invalid pointer upon return.

gottesmm commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -64,16 +95,121 @@ simplifyAndReplace(SILInstruction *inst, CanonicalizeInstruction &pass) {
+    auto extract = cast<SingleValueInstruction>(user);
+    projections.push_back({Projection(extract), extract});
+  }
+  // Avoid removing dead loads from raw SIL. That may weaken definite-init.
```
I wonder if it could also cause problems for debug info. Not sure. May be worth expanding this comment.

atrick replied:
In RAW SIL, I simply don't remove dead loads, regardless of debug info. I'll move this check to the point at which we delete the load so it's independent of whether any projections were rewritten. i.e. the original load will never be deleted in RAW SIL, whether new projection loads were created or not. That way, this pass can't affect diagnostics:
```
_ = deadUseOfInitializedField // Still want an uninitialized use message
```

gottesmm commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -0,0 +1,79 @@
+  pass.newUsers(result);
+  return nextII;
+}
+}
```
Is this an extra '}'

gottesmm commented on lib/SILOptimizer/Mandatory/SILGenCleanup.cpp:
```diff
@@ -0,0 +1,108 @@
+      invalidateAnalysis(&function, invalidKind);
+    }
+  }
+};
```
Extra semiclolon here.

gottesmm reviewed (commented):
Some more suggestions. I went line by line. I need to read the load [copy] a bit more in depth. While you are looking at these I will look in more depth there.

atrick commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -3255,6 +3255,9 @@ class BeginBorrowInst
                              LValue->getType().getObjectType()) {}
+
+public:
+  Operand *getSingleNonEndingUse() const;
```
MG: Plz comment this.

gottesmm commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -103,6 +110,41 @@ simplifyAndReplace(SILInstruction *inst, CanonicalizeInstruction &pass) {
 
+// Replace all uses of an original struct or tuple extract instruction, with the
+// given load instruction. The caller ensures that the load only loads the
+// extracted field.
```
You are assuming below a specific form of the code (e.x.: the cast <BeginBorrowInst>()). Can you add an example in this comment just to be /really/ explicit.

atrick replied:
Sure. Done.

gottesmm commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -103,6 +110,41 @@ simplifyAndReplace(SILInstruction *inst, CanonicalizeInstruction &pass) {
+                          .createBeginBorrow(loadInst->getLoc(), loadInst);
+    pass.notifyNewInstruction(newBorrow);
+
+    for (Operand *use : origBorrow->getUses()) {
```
Do you think this is general enough to make it a helper on begin borrow. I do think that at some point we are going to create a helper to find the end borrows of a borrow scope. Just not sure if find singular end borrow given single use is useful enough.

But it does feel like this type of operation may be common when writing peepholes.

atrick replied:
Ok. I added this because we'll need it for sure.

gottesmm commented on lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp:
```diff
@@ -142,9 +188,26 @@ splitAggregateLoad(LoadInst *loadInst, CanonicalizeInstruction &pass) {
+        destroys.push_back(destroy);
+        continue;
+      } else
+        return nextII;
```
Can you not do an else return here? That is against LLVM style. I think you could re-organize this code to eliminate it like so:

```
if (needsBorrow) {
  if (auto *destroy = dyn_cast<DestroyValueInst>(user)) { ... }
  auto *borrow = dyn_cast<BeginBorrowInst>(user);
  if (!borrow)
    return nextII;
  ...
}
```

atrick replied:
OK. I originally wrote it that way. I put it back.

gottesmm reviewed (approved):
Some small additional asks. But beyond that... LGTM! This is a great PR! Great work Andy!

atrick committed:
Fix eraseFromParentWithdebugInsts to call a callback.

The recursivelyDeleteTriviallyDeadInstructions utility takes a
callBack to be called for every deleted instruction. However, it
wasn't passing this callBack to eraseFromParentWithdebugInsts. The
callback was used to update an iterator in some cases, so not calling
it resulted in iterator invalidation.

Doing this also cleans up the both APIs:
recursivelyDeleteTriviallyDeadInstructions and eraseFromParentWithdebugInsts.

atrick committed:
Add a BeginBorrowInst::getEndBorrows helper.

atrick committed:
Add a comment on replaceUsesOfExtract per review feedback.

atrick committed:
Avoid else-return in splitAggregateLoad.

atrick committed:
Add a CanonicalizeInstruction test case for nontrivial loads.

atrick commented:
@swift-ci test.

atrick commented:
@swift-ci test source compatibility
